src/github/github.installation.service.ts

```diff
@@ -0,0 +1,173 @@
+    const org = await this.allowedOrgRepository.findOne({
+      where: { organizationName, status: OrganizationStatus.ACTIVE },
+    });
+    return !!org;
+  }
+
+  async isUserAllowed(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<boolean> {
+    try {
+      const org = await this.findByName(organizationName);
+      if (org.status !== OrganizationStatus.ACTIVE) {
+        return false;
+      }
+      if (org.allowAllUsers) {
+        return true;
+      }
+      return org.activeUsers.includes(githubUsername);
+    } catch (error) {
+      this.logger.error(`Error checking user allowance: ${error.message}`);
+      return false;
+    }
+  }
+
+  async addAllowedUser(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    if (org.inactiveUsers.includes(githubUsername)) {
+      org.inactiveUsers = org.inactiveUsers.filter(
+        (username) => username !== githubUsername,
+      );
+    }
+    if (!org.activeUsers.includes(githubUsername)) {
+      org.activeUsers.push(githubUsername);
+      return this.allowedOrgRepository.save(org);
+    }
+    return org;
+  }
+
+  async removeAllowedUser(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    org.activeUsers = org.activeUsers.filter(
+      (username) => username !== githubUsername,
+    );
+    if (!org.inactiveUsers.includes(githubUsername)) {
+      org.inactiveUsers.push(githubUsername);
+    }
+    return this.allowedOrgRepository.save(org);
+  }
+
+  async updateStatus(
+    organizationName: string,
+    newStatus: OrganizationStatus,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    org.status = newStatus;
+    return this.allowedOrgRepository.save(org);
+  }
+
+  async uninstall(id: number) {
+    const now = Math.floor(Date.now() / 1000);
+
+    const pemFilename = Env.GITHUB.APP_PEM;
+    const pemFilePath = path.join(__dirname, '../../../', pemFilename);
+    const privateKey = fs.readFileSync(pemFilePath, 'utf8');
+
+    const payload = {
+      iat: now,
+      exp: now + 600,
+      iss: Env.GITHUB.APP_ID,
+      installation_id: id,
+    };
+    const token = jwt.sign(payload, privateKey, { algorithm: 'RS256' });
+    const octokit = new Octokit({
+      auth: token,
+    });
+
+    await octokit.request('DELETE /app/installations/{installation_id}', {
+      installation_id: id,
+      headers: {
+        'X-GitHub-Api-Version': '2022-11-28',
+      },
+    });
+  }
```

Comment:
PullProofAI

The `uninstall` method should include proper error handling for file reading, JWT token generation, Octokit instantiation, and API requests to ensure robustness and reliability.



<!-- This is an auto-generated comment by PullProofAI -->